aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/XrayRulesSampler.java

```diff
@@ -70,4 +96,46 @@ public SamplingResult shouldSample(
   public String getDescription() {
     return "XrayRulesSampler{" + Arrays.toString(ruleAppliers) + "}";
   }
+
+  List<GetSamplingTargetsRequest.SamplingStatisticsDocument> snapshot(Date now) {
+    return Arrays.stream(ruleAppliers)
+        .map(rule -> rule.snapshot(now))
+        .filter(Objects::nonNull)
+        .collect(Collectors.toList());
+  }
+
+  long nextTargetFetchTimeNanos() {
+    return Arrays.stream(ruleAppliers)
+        .mapToLong(SamplingRuleApplier::getNextSnapshotTimeNanos)
+        .min()
+        // There is always at least one rule in practice so this should never be exercised.
+        .orElseGet(() -> System.nanoTime() + AwsXrayRemoteSampler.DEFAULT_TARGET_INTERVAL_NANOS);
+  }
+
+  XrayRulesSampler withTargets(
+      Map<String, SamplingTargetDocument> ruleTargets,
+      Set<String> requestedTargetRuleNames,
+      Date now) {
+    long defaultNextSnapshotTimeNanos =
+        clock.nanoTime() + AwsXrayRemoteSampler.DEFAULT_TARGET_INTERVAL_NANOS;
+    SamplingRuleApplier[] newAppliers =
+        Arrays.stream(ruleAppliers)
+            .map(
+                rule -> {
+                  SamplingTargetDocument target = ruleTargets.get(rule.getRuleName());
```

Comment:
Does this constrain the amount of rule appliers to always be the same as the starting amount of rule appliers? For example, if a customer adds a new rule, and we fetch it, I believe there should now be 1 additional applier (because appliers are 1:1 with customer rules). But in this map, the `newAppliers` is always the same size as the current appliers. Does this get updated elsewhere? And the use I described of adding a new rule will still work with this logic?

Reply:
Rule updates are handled by `AwsXrayRemoteSampler.getAndUpdateSampler`. If rules changed, any pending target fetches are cancelled and they'll be fetched again for the new rules. So basically "updating targets" indeed can never increase the number of rules, but "fetching rules" can and will invalidate targets if rules change.